node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/task/util/process/ChildProcessImpl.java

```diff
@@ -34,49 +45,63 @@ public String commandLine() {
     }
 
     public String getUtf8Output() {
-        waitForTermination();
-        return new UnixPath(processOutputPath).readUtf8File();
+        if (!utf8OutputCache.isPresent()) {
+            waitForTermination();
+            utf8OutputCache = Optional.of(process.getUtf8Output());
+        }
+
+        return utf8OutputCache.get();
     }
 
     public ChildProcessImpl waitForTermination() {
-        while (true) {
-            try {
-                process.waitFor();
-            } catch (InterruptedException e) {
-                // ignoring
-                continue;
-            }
-
-            return this;
-        }
+        process.waitForTermination();
+        return this;
     }
 
     public int exitValue() {
         waitForTermination();
-        return process.exitValue();
+        return process.exitCode();
     }
 
     public ChildProcess throwIfFailed() {
         waitForTermination();
-        if (process.exitValue() != 0) {
-            throw new CommandException("Execution of program [" + commandLine +
-                    "] failed, stdout/stderr was: <" + suffixOfOutputForLog() + ">");
+        int exitCode = process.exitCode();
+        if (exitCode != 0) {
+            String message = debugDescription("terminated with non-zero exit code " + exitCode);
+            throw new CommandException(message);
         }
 
         return this;
     }
 
-    private String suffixOfOutputForLog() {
-        String output = getUtf8Output();
+    @Override
+    public UnexpectedOutputException newUnexpectedOutputException(String problem) {
```

Comment:
Why is this a method? Think this should just be done in the constructor of the exception.

Reply:
The method needs access to commandLine() and getUtf8Output() from this instance. I could make a constructing taking 'problem', and those 2 arguments?

Reply:
Sounds good.

Reply:
Done